### PR TITLE
feat: deno doc --html

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit adds static documentation site generate to "deno doc"
subcommand.

Example:
```
$ deno doc --html --name="My library" ./mod.ts
# outputs to ./docs/

$ deno doc --html --name="My library" --output=./documentation/ ./mod.ts ./file2.js
# outputs to ./documentation/

$ deno doc --html --name="My library" ./**/mod.ts
# generate docs for all files with "mod.ts" name
```

Closes https://github.com/denoland/deno/issues/8233

### CODE REVIEW

File: cli/args/flags.rs [hunk 1]
```diff
@@ -1350,6 +1363,28 @@ Show documentation for runtime built-ins:
...
+            .action(ArgAction::Set)
+            .required_if_eq("html", "true")
+            .require_equals(true)
+        )
+        .arg(
+          Arg::new("output")
+            .long("output")
+            .help("Directory for HTML documentation output")
+            .action(ArgAction::Set)
+            .require_equals(true)
```
Comment (crowlKats) [inline cli/args/flags.rs:1386]:
```suggestion
            .require_equals(true)
            .value_hint(ValueHint::DirPath)
```

File: cli/args/flags.rs [hunk 2]
```diff
@@ -3103,9 +3138,29 @@ fn doc_parse(flags: &mut Flags, matches: &mut ArgMatches) {
...
+    "".to_string()
+  };
+  let output = if html {
+    matches
+      .remove_one::<String>("output")
+      .unwrap_or(String::from("./docs/"))
+  } else {
+    "".to_string()
+  };
+  let html = if html {
```
Comment (crowlKats) [inline cli/args/flags.rs:3154]:
```suggestion
  let html = if matches.get_flag("html") {
    let name = matches.remove_one::<String>("name").unwrap();
    let output = matches
      .remove_one::<String>("output")
      .unwrap_or(String::from("./docs/"));
```

File: cli/tools/doc.rs [hunk 3]
```diff
@@ -85,36 +89,106 @@ pub async fn print_docs(
...
+  let mut output_dir_relative_str =
+    output_dir_relative.to_string_lossy().to_string();
+  output_dir_relative_str = output_dir_relative_str
+    .strip_prefix('/')
+    .unwrap_or(&output_dir_relative_str)
+    .to_string();
+  output_dir_relative_str = output_dir_relative_str
+    .strip_suffix('/')
+    .unwrap_or(&output_dir_relative_str)
+    .to_string();
```
Comment (crowlKats) [inline cli/tools/doc.rs:126]: duplicate?

File: cli/args/flags.rs [hunk 4]
```diff
@@ -1350,6 +1363,28 @@ Show documentation for runtime built-ins:
...
+        .arg(
+          Arg::new("name")
+            .long("name")
+            .help("The name that will be displayed in the docs")
+            .action(ArgAction::Set)
+            .required_if_eq("html", "true")
+            .require_equals(true)
+        )
+        .arg(
+          Arg::new("output")
```
Comment (crowlKats) [inline cli/args/flags.rs:1382]: maybe parse & validate as PathBuf, and store a PathBuf in `DocHtmlFlag` instead of String?

**Comment by bartlomieju** [comment]
Blocked by https://github.com/denoland/deno/pull/21018

File: cli/args/flags.rs [hunk 5]
```diff
@@ -1357,6 +1370,30 @@ Show documentation for runtime built-ins:
             .help("Output documentation in JSON format")
             .action(ArgAction::SetTrue),
         )
+        .arg(
+          Arg::new("html")
+            .long("html")
+            .help("UNSTABLE: Output documentation in HTML format")
```
Comment (bartlomieju) [inline cli/args/flags.rs:1376]: Not sure if it's worth putting `UNSTABLE:` here

File: cli/tests/testdata/doc/use_import_map.out [hunk 6]
```diff
@@ -1,5 +1,5 @@
 Defined in [WILDCARD]/doc/module/fun.js:2:1
 
-function fun(_a, _b)
+function fun(_a, _b): void
```
Comment (bartlomieju) [inline cli/tests/testdata/doc/use_import_map.out:3]: Change caused by `deno_doc` update

File: cli/tools/doc.rs [hunk 7]
```diff
@@ -109,49 +135,101 @@ pub async fn print_docs(
...
+      .with_context(|| format!("Failed to write file {:?}", this_path))?;
   }
+
+  println!(
+    "{}",
+    colors::green(format!(
+      "Written {} files to {:?}",
+      no_of_files, html_options.output
+    ))
+  );
```
Comment (bartlomieju) [inline cli/tools/doc.rs:203]: Suggestions welcome for other prompts

File: cli/tools/doc.rs [hunk 8]
```diff
@@ -109,49 +135,101 @@ pub async fn print_docs(
...
+    let prefix = this_path.parent().with_context(|| {
+      format!("Failed to get parent path for {:?}", this_path)
+    })?;
+    std::fs::create_dir_all(prefix)
+      .with_context(|| format!("Failed to create directory {:?}", prefix))?;
+    std::fs::write(&this_path, content)
+      .with_context(|| format!("Failed to write file {:?}", this_path))?;
   }
+
+  println!(
```
Comment (dsherret) [inline cli/tools/doc.rs:197]: Should this be `log::info!`?

File: cli/tools/doc.rs [hunk 9]
```diff
@@ -37,52 +85,30 @@ pub async fn print_docs(
...
+      )
+      .await?
     }
-    DocSourceFileFlag::Paths(source_files) => {
+    DocSourceFileFlag::Paths(ref source_files) => {
       let module_graph_builder = factory.module_graph_builder().await?;
       let maybe_lockfile = factory.maybe_lockfile();
 
+      let expanded_globs =
+        expand_globs(source_files.iter().map(PathBuf::from).collect())?;
```
Comment (dsherret) [inline cli/tools/doc.rs:103]: I'm not sure about doing globbing here. Do we support globs in any other arguments? Maybe we should move this to a separate PR.

**Review by dsherret — APPROVED** [review]
LGTM

File: cli/args/flags.rs (hunk 5 above)
Comment (crowlKats) [reply cli/args/flags.rs:1376]: no, dont think so

File: cli/tools/doc.rs (hunk 9 above)
Comment (bartlomieju) [reply cli/tools/doc.rs:103]: Yes, globs are supported in other subcommands as well

File: cli/tools/doc.rs (hunk 8 above)
Comment (bartlomieju) [reply cli/tools/doc.rs]: Using `log::info!` seems to not print this info by default. I'm gonna revert it
